### PR TITLE
feat: inline actor bundles

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:

- Inline actor bundles
- Re-enable state migration unit tests

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2765

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: actor_bundles/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+*
```
Comment (LesnyRumcajs) [inline actor_bundles/.gitignore:1]: Could we just add `actor_bundles` to the root `.gitignore`? In general, it's a bit surprising to suddenly have such directory. Maybe we could use `target/`?

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
We need to make it less of a burden for developers. This approach forces builds even if there are no changes.

That is to say:
```
❯ cargo build
   Compiling forest-filecoin v0.11.1 (/home/rumcajs/prj/forest2)
    Finished dev [unoptimized] target(s) in 15.04s

forest2 on  hm/inline-actor-bundles [$?] is 📦 v0.11.1 via  v18.16.1 via 🦀 v1.71.0 took 15s
❯ cargo build
   Compiling forest-filecoin v0.11.1 (/home/rumcajs/prj/forest2)
    Finished dev [unoptimized] target(s) in 15.13s
```

as opposed to
```
❯ cargo build
   Compiling forest-filecoin v0.11.1 (/home/rumcajs/prj/forest)
    Finished dev [unoptimized] target(s) in 20.02s

forest on  query-cid-2 [$!?] is 📦 v0.11.1 via  v18.16.1 via 🦀 v1.71.0 took 20s
❯ cargo build
    Finished dev [unoptimized] target(s) in 0.15s
```

**Comment by hanabi1224** [comment]
> This approach forces builds even if there are no changes.

@LesnyRumcajs Fixed by tweaking `cargo:rerun-if-changed` lines, it should now only rebuild when bundle car cache is changed

File: actor_bundles/.gitignore (hunk 1 above)
Comment (hanabi1224) [reply actor_bundles/.gitignore]: Moved to `target/actor_bundles/`

File: build.rs [hunk 2]
```diff
@@ -1,26 +1,213 @@
...
+        car_readers
+            .iter()
+            .flat_map(|it| it.header.roots.iter())
+            .unique()
+            .cloned()
+            .collect::<Vec<_>>(),
+    );
+
+    let mut zstd_encoder = ZstdEncoder::with_quality(
+        async_fs::File::create(Path::new(&env::var("OUT_DIR")?).join("actor_bundles.car.zst"))
```
Comment (LesnyRumcajs) [inline build.rs:110]: This is also defined in `src/daemon/bundle.rs`. Let's have a single source of truth.

File: build.rs [hunk 3]
```diff
@@ -1,26 +1,213 @@
...
+use once_cell::sync::Lazy;
 use protobuf_codegen::Customize;
+use tokio::task::JoinSet;
 use walkdir::WalkDir;
 
 const PROTO_DIR: &str = "proto";
 const CARGO_OUT_DIR: &str = "proto";
 
-fn main() -> anyhow::Result<()> {
+const ACTOR_BUNDLE_CACHE_DIR: &str = "target/actor_bundles/";
```
Comment (LesnyRumcajs) [inline build.rs:27]: It'd be great to have it as a `PathBuf`. The less we operate on strings and more on strongly typed objects, the less likely we'll introduce a bug.

File: src/daemon/bundle.rs [hunk 4]
```diff
@@ -1,84 +1,28 @@
...
+    use super::*;
 
-    // If the bundle already exists, return a reader to it.
-    if bundle_path.exists() {
-        let file = async_fs::File::open(bundle_path).await?;
-        return Ok(futures::io::BufReader::new(file));
+    #[tokio::test]
+    async fn test_load_actor_bundles() {
+        let db = fvm_ipld_blockstore::MemoryBlockstore::new();
+        load_actor_bundles(&db).await.unwrap();
```
Comment (LesnyRumcajs) [inline src/daemon/bundle.rs:26]: Let's add some assertions here, even with hard-coded Cids.

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
Let's also remove the logic from places like https://github.com/ChainSafe/forest/blob/hm/inline-actor-bundles/src/networks/devnet/mod.rs#L112 . To my understanding, this is no longer needed.

File: build.rs (hunk 3 above)
Comment (hanabi1224) [reply build.rs]: Fixed

File: src/daemon/bundle.rs (hunk 4 above)
Comment (hanabi1224) [reply src/daemon/bundle.rs]: Fixed

File: build.rs (hunk 2 above)
Comment (hanabi1224) [reply build.rs:87]: Not sure how to share the code, `include_bytes!(concat!(env!("OUT_DIR"), "/actor_bundles.car.zst"));` requires str literals and is evaluated at compile time

Comment (hanabi1224) [reply build.rs:87]: Below code does not compile, only literals (like `"foo"`, `-42` and `3.14`) can be passed to `concat!()`
```rust
const BUNDLE_NAME: &str = "/actor_bundles.car.zst";
pub const ACTOR_BUNDLES_CAR_ZST: &[u8] = include_bytes!(concat!(env!("OUT_DIR"), BUNDLE_NAME));
```

**Comment by hanabi1224** [comment]
> Let's also remove the logic from places like https://github.com/ChainSafe/forest/blob/hm/inline-actor-bundles/src/networks/devnet/mod.rs#L112 . To my understanding, this is no longer needed.

@LesnyRumcajs  Fixed

File: src/mod.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+use lazy_static::lazy_static;
+use reqwest::Url;
+
+#[derive(Debug)]
+pub struct ActorBundleInfo {
+    pub manifest: Cid,
+    pub url: Url,
+}
+
+lazy_static! {
```
Comment (LesnyRumcajs) [inline src/mod.rs:16]: Perhaps we could avoid this macro with `once_cell::sync::Lazy`

File: src/networks/calibnet/mod.rs [hunk 6]
```diff
@@ -123,26 +121,17 @@ pub static ref HEIGHT_INFOS: [HeightInfo; 21] = [
     HeightInfo {
         height: Height::Shark,
         epoch: 16_800,
-        bundle: Some(ActorBundleInfo {
-            manifest: Cid::try_from("bafy2bzacedbedgynklc4dgpyxippkxmba2mgtw7ecntoneclsvvl4klqwuyyy").unwrap(),
-            url: Url::parse("https://github.com/filecoin-project/builtin-actors/releases/download/v9.0.3/builtin-actors-calibrationnet.car").unwrap()
-        })
+        bundle: Some(Cid::try_from("bafy2bzacedbedgynklc4dgpyxippkxmba2mgtw7ecntoneclsvvl4klqwuyyy").unwrap())
```
Comment (LesnyRumcajs) [inline src/networks/calibnet/mod.rs:124]: Do we still need this? I mean, we have this defined here and in `src/mod.rs`

Comment (hanabi1224) [reply src/networks/calibnet/mod.rs:124]: We still need to map the cid to the network and height. Moving everything to mod.rs might be too much

File: src/mod.rs (hunk 5 above)
Comment (hanabi1224) [reply src/mod.rs:16]: Fixed

**Review by lemmih — APPROVED** [review]

**Pull request merged**